tests/gradients/core/test_pulse_gradient.py

```diff
@@ -264,6 +265,459 @@ def test_with_parametrized_evolution(self):
+
+        np.random.seed(3751)
+
+        cjac_shape = (num_split_times,) + par_shape
+        if multi_term > 1:
+            cjacs = tuple(np.random.random(cjac_shape) for _ in range(multi_term))
+            psr_coeffs = tuple(np.random.random(num_shifts) for _ in range(multi_term))
+        else:
+            cjacs = np.random.random(cjac_shape)
+            psr_coeffs = np.random.random(num_shifts)
+
+        results_shape = (num_shifts * multi_term, (num_split_times + 2)) + meas_shape
+        new_results_shape = (
+            multi_term,
+            num_shifts,
+            num_split_times + 2,
+        ) + meas_shape
+        results = np.random.random(results_shape)
+
+        prefactor = 0.3214
+
+        res = _parshift_and_integrate(
+            results,
+            cjacs,
+            prefactor,
+            psr_coeffs,
+            single_measure=True,
+            has_partitioned_shots=False,
+            use_broadcasting=True,
+        )
+
+        assert isinstance(res, jnp.ndarray)
+        assert res.shape == meas_shape + par_shape
+
+        _results = np.reshape(results, new_results_shape)
+        _cjacs = np.stack(cjacs).reshape((multi_term,) + cjac_shape)
+        _psr_coeffs = np.stack(psr_coeffs).reshape((multi_term, num_shifts))
+        meas_letter = "" if meas_shape == () else "a"
+        # Slice away excess results
+        _results = _results[:, :, 1:-1]
+        # With broadcasting, the axes of different shifts and splitting times are
+        # switched for the results tensor, compared to without broadcasting.
+        contraction = f"ms,mst{meas_letter},mt...->{meas_letter}..."
+        expected = np.einsum(contraction, _psr_coeffs, _results, _cjacs)
+        assert np.allclose(res, expected * prefactor)
+
+    @pytest.mark.parametrize("multi_term", [1, 4])
+    @pytest.mark.parametrize("meas_shape", [(), (4,)])
+    @pytest.mark.parametrize("par_shape", [(), (3,), (2, 2)])
+    @pytest.mark.parametrize("num_shifts", [2, 5])
+    @pytest.mark.parametrize("num_split_times", [1, 3])
+    def test_multi_measure_or_multi_shots(
+        self, num_split_times, num_shifts, par_shape, multi_term, meas_shape
+    ):
+        """Test that ``_parshift_and_integrate`` works with results for multiple measurements
+        per shift and splitting time and with a single setting of shots, or alternatively with
+        a single measurement but multiple shot settings. This corresponds to
+        ``single_measure=False and has_partitioned_shots=False`` or
+        ``single_measure=True and has_partitioned_shots=True``. The test is parametrized with whether
+        or not there are multiple Hamiltonian terms to take into account (and sum their
+        contributions), with the shape of the single measurement and of the parameter, with
+        the number of shifts in the shift rule and with the number of splitting times.
+        """
+        from jax import numpy as jnp
+
+        np.random.seed(3751)
```

Comment:
Could set the seed only once in the test file?

Reply:
Wouldn't this make the test shaky w.r.t randomized order of test execution? :thinking:

Reply:
Yes also since jax and non-jax tests are executed separately. Could we also use `jax.random` for jax tests?

Reply:
If there is a preference for that, sure :)